lsコマンド4を作成

samovar1214 opened the pull request.

samovar1214 committed:
lsコマンド4を作成

yuuu commented on 04.ls/ls.rb:
```diff
@@ -26,13 +27,63 @@ def print_rows(rows)
+end
+
+def print_long_format(filenames)
+  file_infos = filenames.map { |filename| [filename, File.lstat(filename)] }
```
このままでもOKですが、 `it` を使うとさらにスッキリ書くことができます 👍 

```ruby
file_infos = filenames.map { [it, File.lstat(it)] }
```

samovar1214 replied:
これならブロック引数を明示的に書かなくても参照されるんですね！✨`it`を使った書き方に修正します。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -26,13 +27,63 @@ def print_rows(rows)
+  puts "total #{total_blocks}"
+
+  file_infos.each do |filename, file_info|
+    printf(
```
ここは `printf` でなければいけないのは `size` くらいで、 `size` のみ4桁分の空白を上手く入れられれば `printf` でなくてもよいですね。

`printf` を使わない方が全体のコードの記述量を少なくできるはずなので、リファクタリングをお願いします 🙏

samovar1214 replied:
たしかにここの`printf`は`size`の桁数を揃えるためだけに頑張っている状態ですね…。`printf`以外で同様の表示がされるようリファクタリングしてみます🫡

yuuu commented on 04.ls/ls.rb:
```diff
@@ -26,13 +27,63 @@ def print_rows(rows)
+if options['l']
+  print_long_format(filenames)
+else
+  column_count = 3
```
`print_long_format` と同様にここも1つのメソッドを呼ぶようにしたいですね。対称性のあるコードは読みやすさにつながるためです。

samovar1214 replied:
対称性のあるコード、勉強になります…！`if`の分岐を「どっちの表示メソッドを呼ぶか」だけにして、読みやすくなるように修正します💪🏻

samovar1214 committed:
-lの表示処理とメソッド呼び出しを修正

重複している変数名（filename）をitに置き換える書き方に修正し、printfからputを使った表示方法に変更することで記述量を少なくしました。また、-lオプションを指定した場合とそうでない場合でそれぞれ1つのメソッドを呼ぶようにして、対称性を持たせました。

yuuu commented on 04.ls/ls.rb:
```diff
@@ -54,22 +60,20 @@ def render_filetype_and_filemode(file_info)
-      mtime: file_info.mtime.strftime('%b %e %H:%M'),
-      name: filename
-    )
+    mode = render_filetype_and_filemode(file_info)
```
ローカル変数の数も減らせるのであればそれに越したことはないので次のように文字列配列→joinをすると読みやすくなりそうです 👍 

```ruby
puts [
  render_filetype_and_filemode(file_info),
  file_info.nlink,
  # ...
].join(' ')
```

samovar1214 replied:
たしかに直接配列に入れてjoin(" ") で並べちゃえば、ローカル変数を減らせますし読みやすくなりますね…！
示していただいた方法で書いてみます👀

samovar1214 committed:
-lの表示処理を配列からjoinする形に修正

ローカル変数を使わずに項目をそのまま配列に入れてjoinで並べることで、可読性が向上するように修正しました。
